sources/web/datalab/fileSearch.ts

```diff
@@ -0,0 +1,132 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/// <reference path="../../../third_party/externs/ts/node/node.d.ts" />
+/// <reference path="../../../third_party/externs/ts/chokidar/chokidar.d.ts" />
+/// <reference path="common.d.ts" />
+
+import http = require('http');
+import url = require('url');
+import fs = require('fs');
+import path = require('path');
+import logging = require('./logging');
+import chokidar = require('chokidar');
+
+let appSettings: common.Settings;
+let fileIndex: string[] = [];
+let indexReady: boolean = false;
+let tooManyFiles: boolean = false;
+const fileCountLimit = 1000000;
+
+// this is matched by the client javascript to display a message that
+// there are more results than shown. Make sure these are in sync
+const clientResultSize = 20; 
+
+/**
+ * Implements the file search/filter request handling.
+ * @param request the incoming search request.
+ * @param response the outgoing search response.
+ */
+function requestHandler(request: http.ServerRequest, response: http.ServerResponse): void {
+  const parsedUrl = url.parse(request.url, true);
+  const pattern = parsedUrl.query['pattern'];  
+  const statusCheck = parsedUrl.query['status'];
+
+  response.writeHead(200, { 'Content-Type': 'application/json' });
+  let results: string[] = [];
+  if (pattern) {
+    results = filter(pattern, fileIndex);
+  }
+  response.write(JSON.stringify({
+    files: results.slice(0, clientResultSize),
+    fullResultSize: results.length,
+    tooManyFiles: tooManyFiles,
+    indexReady: indexReady,
+    indexingEnabled: appSettings.enableFilesystemIndex,
+  }));
+  response.end();
+}
+
+/**
+ * Builds an index of all files in the content directory to make search faster
+ */
+export function indexFiles(): void {
+  const startTime = process.hrtime();
+  logging.getLogger().info('Started indexing file system');
+  chokidar.watch(appSettings.contentDir + '/', {
+      usePolling: true,
+      interval: 1000,               // we don't need high frequency polling
+      ignored: /(^|[\/\\])\../,     // ignore dot files/dirs
+      ignorePermissionErrors: true, // ignore files with no permissions
+    })
+    .on('add', (addedPath) => {
+      if (fileIndex.length >= fileCountLimit) {
+        tooManyFiles = true;
+      } else {
+        fileIndex.push(addedPath.substr(appSettings.contentDir.length + 1));
+      }
+    })
+    .on('unlink', (deletedPath) => {
+      deletedPath = deletedPath.substr(appSettings.contentDir.length + 1);
+      let pos = fileIndex.indexOf(deletedPath);
+      if (pos > -1) {
+        fileIndex.splice(pos, 1);
+      }
+    })
+    .on('ready', () => {
+      indexReady = true;
+      const indexTime = process.hrtime(startTime);
+      logging.getLogger().info('Finished indexing ' + fileIndex.length + ' files in ' + indexTime[0] + ' seconds');
+    })
+    .on('raw', () => {
+      if (!indexReady) {
```

Comment:
Why do you test indexReady here, but not for on('ready')?
Based on the code, I can make guesses as to which of these message types might be received more than once, but it might be nice to add some comments to that effect. Then maybe I wouldn't have to ask the above question.